trin-core/src/portalnet/overlay_service.rs

```diff
@@ -827,23 +827,116 @@ impl OverlayService {
     }
 
     /// Processes a Nodes response.
-    fn process_nodes(&mut self, _nodes: Nodes, source: Enr) {
+    fn process_nodes(&mut self, nodes: Nodes, source: Enr) {
         debug!(
             "[{:?}] Processing Nodes response from node. Node: {}",
             self.protocol,
             source.node_id()
         );
-        // TODO: Implement processing logic.
+        self.process_discovered_enrs(
+            nodes
+                .enrs
+                .into_iter()
+                .map(|ssz_enr| ssz_enr.into())
+                .collect(),
+        );
     }
 
     /// Processes a Content response.
-    fn process_content(&mut self, _content: Content, source: Enr) {
+    fn process_content(&mut self, content: Content, source: Enr) {
         debug!(
             "[{:?}] Processing Content response from node. Node: {}",
             self.protocol,
             source.node_id()
         );
-        // TODO: Implement processing logic.
+        if let Content::Enrs(enrs) = content {
+            self.process_discovered_enrs(enrs.into_iter().map(|ssz_enr| ssz_enr.into()).collect());
+        }
+    }
+
+    /// Processes a collection of discovered nodes.
+    fn process_discovered_enrs(&mut self, enrs: Vec<Enr>) {
+        let local_node_id = self.local_enr().node_id();
+
+        // Acquire write lock here so that we can perform node lookup and insert/update atomically.
+        // Once we acquire the write lock for the routing table, there are no other locks that we
+        // need to acquire, so we should not create a deadlock.
+        let mut kbuckets = self.kbuckets.write();
+
+        for enr in enrs {
+            let node_id = enr.node_id();
+
+            // Ignore ourself.
+            if node_id == local_node_id {
+                continue;
+            }
+
+            let key = kbucket::Key::from(node_id);
+            let node = match kbuckets.entry(&key) {
+                kbucket::Entry::Present(entry, _) => Some(entry.value().clone()),
+                kbucket::Entry::Pending(ref mut entry, _) => Some(entry.value().clone()),
```

Comment:
It's been some time since I originally looked at this, but taking another look at the [bucket update function](https://github.com/sigp/discv5/blob/9b21ec1e274138661fd2fa35bef54143099f2ef2/src/kbucket/bucket.rs#L482), it looks like it is defined to handle both present and pending entries.

I do think that the `present_node` variable name below is a bit confusing, given that the node may be present or pending, so I can change that to something more neutral.